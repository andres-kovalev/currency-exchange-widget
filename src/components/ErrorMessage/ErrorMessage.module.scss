.container {
    background-color: red;
}

.message {
    display: flex;
    align-items: center;
    font-size: 0.5rem;
}

.icon {
    font-size: 0.5rem;
}

.errorIcon {
    margin-right: 0.25rem;
}
