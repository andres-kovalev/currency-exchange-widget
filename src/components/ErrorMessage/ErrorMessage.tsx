import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { Snackbar, SnackbarContent, IconButton } from '@material-ui/core';
import { Error, Close } from '@material-ui/icons';

import styles from './ErrorMessage.module.scss';

type VerticalAlignment = 'top' | 'bottom';
type HorizontalAlignment = 'left' | 'right' | 'center';

interface SnackbarAnchorOrigin {
    vertical: VerticalAlignment;
    horizontal: HorizontalAlignment;
}

const anchorOrigin: SnackbarAnchorOrigin = {
    vertical: 'top',
    horizontal: 'center'
};

interface ErrorMessageProps {
    open?: boolean;
    autoHideDuration?: number;
    message: string;
    onClose: () => void;
}

const ErrorMessage: React.FC<ErrorMessageProps> = ({
    open = false,
    autoHideDuration = 3000,
    message, onClose
}) => (
    <Snackbar
        anchorOrigin={ anchorOrigin }
        open={open}
        autoHideDuration={autoHideDuration}
        onClose={onClose}
    >
        <SnackbarContent
            className={styles.container}
            aria-describedby="error-message"
            message={
                <span id="error-message" className={styles.message}>
                    <Error className={cx(styles.icon, styles.errorIcon)} />
                    { message }
                </span>
            }
            action={[
                <IconButton
                    key="close"
                    aria-label="close"
                    color="inherit"
                    onClick={ onClose }
                >
                    <Close className={styles.icon} />
                </IconButton>
            ]}
        />
    </Snackbar>
);

ErrorMessage.propTypes = {
    open: PropTypes.bool,
    autoHideDuration: PropTypes.number,
    message: PropTypes.string.isRequired,
    onClose: PropTypes.func.isRequired
};

export default ErrorMessage;
