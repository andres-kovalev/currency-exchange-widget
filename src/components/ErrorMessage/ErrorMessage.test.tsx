import React from 'react';
import { shallow, mount, ShallowWrapper } from 'enzyme';
import { Snackbar, SnackbarContent, IconButton } from '@material-ui/core';

import ErrorMessage from './ErrorMessage';

// eslint-disable-next-line @typescript-eslint/no-empty-function
const noop = (): void => {};

describe('ErrorMessage', () => {
    describe('by default', () => {
        let wrapper: ShallowWrapper;

        beforeEach(() => {
            wrapper = shallow(
                <ErrorMessage
                    message="Test message"
                    onClose={ noop }
                />
            );
        });

        it('should render as expected', () => {
            expect(wrapper).toMatchSnapshot();
        });

        it('should render itself using Snackbar component', () => {
            expect(wrapper.find(Snackbar)).toExist();
            expect(wrapper.find(SnackbarContent)).toExist();
        });

        it('should render itself not opened', () => {
            const snackbar = wrapper.find(Snackbar);

            expect(snackbar).toHaveProp({ open: false });
        });
    });

    it('should render itself with provided message', () => {
        const wrapper = mount(
            <ErrorMessage
                open
                message="Test message"
                onClose={ noop }
            />
        );
        const messageSpan = wrapper.find('#error-message');

        expect(messageSpan).toHaveText('Test message');
    });

    it('should call onClose callback on close button click', () => {
        const onClose = jest.fn();

        const wrapper = mount(
            <ErrorMessage
                open
                message="Test message"
                onClose={ onClose }
            />
        );
        const closeButton = wrapper.find(IconButton);

        expect(onClose).not.toHaveBeenCalled();

        closeButton.simulate('click');

        expect(onClose).toHaveBeenCalled();
    });
});
