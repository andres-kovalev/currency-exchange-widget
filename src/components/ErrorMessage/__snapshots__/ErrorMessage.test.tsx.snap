// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorMessage by default should render as expected 1`] = `
<WithStyles(ForwardRef(Snackbar))
  anchorOrigin={
    Object {
      "horizontal": "center",
      "vertical": "top",
    }
  }
  autoHideDuration={3000}
  onClose={[Function]}
  open={false}
>
  <WithStyles(ForwardRef(SnackbarContent))
    action={
      Array [
        <ForwardRef(WithStyles)
          aria-label="close"
          color="inherit"
          onClick={[Function]}
        >
          <UNDEFINED
            className="icon"
          />
        </ForwardRef(WithStyles)>,
      ]
    }
    aria-describedby="error-message"
    className="container"
    message={
      <span
        className="message"
        id="error-message"
      >
        <UNDEFINED
          className="icon errorIcon"
        />
        Test message
      </span>
    }
  />
</WithStyles(ForwardRef(Snackbar))>
`;
